src/script_opt/GenRDs.cc

```diff
@@ -0,0 +1,1286 @@
+		mgr.SetPreFromPre(index, a);
+		mgr.SetPreFromPre(rhs, a);
+
+		if ( aggr->Tag() == EXPR_NAME )
+			{
+			// Don't treat this as an initialization of the
+			// aggregate, since what's changing is instead
+			// an element of it.
+			}
+		else
+			aggr->Traverse(this);
+
+		index->Traverse(this);
+		rhs->Traverse(this);
+
+		return TC_ABORTSTMT;
+		}
+
+	case EXPR_FIELD_LHS_ASSIGN:
+		{
+		auto f = e->AsFieldLHSAssignExpr();
+		auto aggr = f->Op1();
+		auto r = f->Op2();
+
+		mgr.SetPreFromPre(aggr, e);
+		mgr.SetPreFromPre(r, e);
+
+		if ( aggr->Tag() == EXPR_NAME )
+			{
+			// Don't treat as an initialization of the aggregate.
+			}
+		else
+			aggr->Traverse(this);
+
+		r->Traverse(this);
+
+		auto r_def = mgr.GetExprDI(aggr);
+		if ( ! r_def )
+			// This should have already generated a complaint.
+			// Avoid cascade.
+			break;
+
+		auto offset = f->Field();
+		auto field_rd = r_def->FindField(offset);
+
+		if ( ! field_rd )
+			field_rd = r_def->CreateField(offset, f->GetType());
+
+		CreateInitPostDef(field_rd, DefinitionPoint(e), false, r);
+
+		return TC_ABORTSTMT;
+		}
+
+	case EXPR_FIELD:
+		{
+		auto f = e->AsFieldExpr();
+		auto r = f->Op();
+
+		mgr.SetPreFromPre(r, e);
+
+		if ( r->Tag() != EXPR_NAME && r->Tag() != EXPR_FIELD )
+			break;
+
+		if ( analysis_options.usage_issues > 1 )
+			{
+			auto r_def = mgr.GetExprDI(r);
+
+			if ( r_def && ! r_def->RootID()->GetAttr(ATTR_IS_SET) )
+				{
+				auto fn = f->FieldName();
+				auto field_rd = mgr.GetConstID_DI(r_def.get(), fn);
+
+				auto e_pre = mgr.GetPreMinRDs(e);
+				if ( ! field_rd || ! e_pre->HasDI(field_rd) )
+					printf("record field possibly used without being set: %s\n", obj_desc(e).c_str());
+				}
+			}
+
+		if ( r->Tag() == EXPR_NAME )
+			{
+			auto r_id = r->AsNameExpr()->Id();
+			if ( r_id->IsGlobal() )
+				// Don't worry about record fields in globals.
+				return TC_ABORTSTMT;
+
+			// For names, we care about checking the name
+			// itself, but if it's a record we don't want to
+			// complain about missing fields, as they're
+			// irrelevant other than the one specifically
+			// being referenced.  So we do the CheckVar here
+			// and don't descend recursively.
+			CheckVar(r, r_id, false);
+			}
+
+		else
+			// Recursively check the subexpression.
+			r->Traverse(this);
+
+		return TC_ABORTSTMT;
+		}
+
+	case EXPR_HAS_FIELD:
+		{
+		auto hf = e->AsHasFieldExpr();
+		auto r = hf->Op();
+
+		mgr.SetPreFromPre(r, e);
+
+		// Treat this as a definition of r$fn, since it's
+		// ensuring that that field exists.  That's not quite
+		// right, since this expression's parent could be a
+		// negation, but at least we know that the script
+		// writer is thinking about whether it's defined.
+
+		if ( r->Tag() == EXPR_NAME )
+			{
+			auto id_e = r->AsNameExpr();
+			auto id = id_e->Id();
+			auto id_rt = id_e->GetType()->AsRecordType();
+			auto id_di = mgr.GetID_DI(id);
+
+			if ( ! id_di /*### && ! analysis_options.inliner */ )
```

Comment:
Thanks for flagging, let me look into this.  I know there are set-but-not-used issues that inlining can create (because of the opportunity to do constant propagation across the inlining, and so prune some branches that have the actual uses), but am not remembering offhand if there's something similar for used-but-not-set.  Will follow up once I've tried this out.

Reply:
This turned out to be fruitful to revisit.  In the process I found a number of bugs.  I've now resolved those, and also renamed the attribute to `&is_assigned` per discussion.  Back over to you, Jon.